### PR TITLE
[pilot, do not merge] Artifacts pilot branch

### PR DESCRIPTION
**Pull request by shawnlewis**
This PR contains support for W&B Artifacts.

### CODE REVIEW

File: wandb/wandb_run.py [hunk 1]
```diff
@@ -203,9 +205,9 @@ def send_message(self, options):
         """ Sends a message to the wandb process changing the policy
         of saved files.  This is primarily used internally by wandb.save
         """
-        if not options.get("save_policy") and not options.get("tensorboard"):
+        if not options.get("save_policy") and not options.get("tensorboard") and not options.get("log_artifact") and not options.get("use_artifact"):
             raise ValueError(
-                "Only configuring save_policy and tensorboard is supported")
+                "Only configuring save_policy, tensorboard  and log_artifact is supported")
```
Comment (raubitsj) [inline wandb/wandb_run.py:210]: nit: and use_artifact

File: standalone_tests/artifact_storage.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import sys
+import tempfile
+
+import wandb
+
+def artifact_with_various_paths():
+    art = wandb.Artifact(type='artsy', name='my artys')
+
+    # internal file
+    with open('random.txt', 'w') as f:
```
Comment (raubitsj) [inline standalone_tests/artifact_storage.py:12]: nit: might want to put this all in a artifact_storage directory just to keep this a bit cleaner.

File: wandb/apis/public.py [hunk 3]
```diff
@@ -489,6 +607,18 @@ def path(self):
     def __repr__(self):
         return "<Project {}>".format("/".join(self.path))
 
+    @normalize_exceptions
+    def artifacts(self, per_page=50):
```
Comment (shawnlewis) [inline wandb/apis/public.py:611]: This should be artifact_types.

File: wandb/apis/public.py [hunk 4]
```diff
@@ -1141,9 +1275,9 @@ class File(object):
     def __init__(self, client, attrs):
         self.client = client
         self._attrs = attrs
-        if self.size == 0:
-            raise AttributeError(
-                "File {} does not exist.".format(self._attrs["name"]))
+        #if self.size == 0:
```
Comment (shawnlewis) [inline wandb/apis/public.py:1278]: I think we may be able to get rid of all changes to the File object here.

File: wandb/apis/public.py [hunk 5]
```diff
@@ -1538,3 +1678,513 @@ def _load_next(self):
...
+        return self._attrs
+
+    # The only file should be wandb_manifest.json
+    def _files(self, names=None, per_page=50):
+        return ArtifactFiles(self.client, self, names, per_page)
+
+    def _load_manifest(self):
+        if self._manifest is None:
+            index_file_url = self._attrs['currentManifest']['file']['url']
+            print(index_file_url)
```
Comment (shawnlewis) [inline wandb/apis/public.py:2124]: Delete

File: wandb/util.py [hunk 6]
```diff
@@ -839,6 +857,18 @@ def guess_data_type(shape, risky=False):
     return None
 
 
+def download_file_from_url(dest_path, source_url, api_key=None):
```
Comment (shawnlewis) [inline wandb/util.py:860]: can probably get rid of this if we can get rid of changes to File in public API

File: wandb/wandb_run.py [hunk 7]
```diff
@@ -406,6 +414,69 @@ def save(self, id=None, program=None, summary_metrics=None, num_retries=None, ap
...
+                        'aliases': aliases
+                    }
+                })
+            elif isinstance(artifact, ApiArtifact):
+                self.api.use_artifact(artifact.id)
+                return artifact
+            else:
+                # TODO: Improve message with instructions, maybe share base class with wandb.Artifact, and the
+                # API artifact?
+                raise ValueError('You must pass an instance of wandb.Artifact, or wandb.Api().artifact() to use_artifact')
```
Comment (shawnlewis) [inline wandb/wandb_run.py:453]: Update this message and remove TODO

**Pull request merged**